### PR TITLE
Various fixes and extensions for smart scripts:

### PR DESCRIPTION
**Pull request by ghost**
- Possible crashes fixed
- Memory leak fixed
- Implemented checking of vehicle conditions
- Extended eventphasemask to 12 bits (sql required to change DB field type)
- SMART_EVENT_GOSSIP_HELLO - added possibility to detect for gameobject reportUse call
- Renamed action SMART_ACTION_SET_FLY to SMART_ACTION_SET_DISABLE_GRAVITY (to reflect actual functionality)
- Added targetsLimit to action SMART_ACTION_CAST and SMART_ACTION_INVOKER_CAST to limit max amount of targets (selected randomly)
- Action SMART_ACTION_TALK corrected to always work as intended
- Properly call GroupEventHappens in action SMART_ACTION_CALL_GROUPEVENTHAPPENS if invoker was charmed or owned by the player
- Properly utilize followAngle in action SMART_ACTION_FOLLOW (db orientation should be in degrees), but keep backward compatibility
- Added action SMART_ACTION_SET_CAN_FLY (119) 0/1
- Added action SMART_ACTION_REMOVE_AURAS_BY_TYPE (120) AuraType, can be used to exit vehicle for example
- Added action SMART_ACTION_SET_SIGHT_DIST (121) sightDist
- Added action SMART_ACTION_FLEE (122) fleeTime
- Added action SMART_ACTION_ADD_THREAT (123) +threat, -threat
- Added action SMART_ACTION_LOAD_EQUIPMENT (124) equipmentId
- Added action SMART_ACTION_TRIGGER_RANDOM_TIMED_EVENT (125) minId, maxId
- Added action SMART_ACTION_REMOVE_ALL_GAMEOBJECTS (126), removes all owned gameobjects
- Added action SMART_ACTION_STOP_MOTION (127), stopMoving, movementExpired
- Extended target SMART_TARGET_HOSTILE_SECOND_AGGRO with following parameters maxdist, playerOnly, powerType + 1
- Extended target SMART_TARGET_HOSTILE_LAST_AGGRO with following parameters maxdist, playerOnly, powerType + 1
- Extended target SMART_TARGET_HOSTILE_RANDOM with following parameters maxdist, playerOnly, powerType + 1
- Extended target SMART_TARGET_HOSTILE_RANDOM_NOT_TOP with following parameters maxdist, playerOnly, powerType + 1
- Extended target SMART_TARGET_THREAT_LIST with maxdist
- Extended target SMART_TARGET_OWNER_OR_SUMMONER to be able to get charmer/owner of current owner
- Added new target SMART_TARGET_FARTHEST with maxDist, playerOnly, isInLos restrictions
- Added SpellHit hook to GameObjectAI and extended SmartGameObjectAI to call SMART_EVENT_SPELLHIT when gameobject is hit by spell
- Call GameObjectAI Reset hook on gameobject respawn (for ex. to reset one time events in smart scripts)
- Fixed some logic errors in code
- SmartAI Escorts properly despawn escortee if no players are in range
- Disable Evading while charming creature with SmartAI
- Don't call SMART_EVENT_RESPAWN for dead units before they actually respawn
- Don't call SMART_EVENT_RESPAWN for not spawned gameobjects
- Properly call SMART_EVENT_RESPAWN for gameobject respawn
- Allow action SMART_ACTION_SET_IN_COMBAT_WITH_ZONE to utilize targetlist
- Allow action SMART_ACTION_CALL_FOR_HELP to utilize targetList
- Allow action SMART_ACTION_SET_INVINCIBILITY_HP_LEVEL to utilize targetList
- Allow action SMART_ACTION_SET_VISIBILITY to utilize targetList
- Allow action SMART_ACTION_SET_ACTIVE to utilize targetList
- Allow action SMART_ACTION_ATTACK_START to select random attack target instead of first on the list
- Allow gameobjects to summon gameobjects with action SMART_ACTION_SUMMON_GO
- Properly store action invokers for action SMART_ACTION_WP_START, if no player invokers are found, distance despawn check won't be used
- Allow action SMART_ACTION_WP_RESUME to compensate for the state the unit actually is in (eg. combat)
- Allow action SMART_ACTION_MOVE_TO_POS to select random of the avaiable targets, not only the first one.
- Allow action SMART_ACTION_MOVE_TO_POS to utilize x, y, z parameters as an offset to calculated coordinates
- Action SMART_ACTION_RESPAWN_TARGET should never modify respawntime of already spawned gameobjects, use dedicated function
- Properly delete ontime events created by SMART_ACTION_CREATE_TIMED_EVENT
- If action could not be started because conditions were not satisfied, do not recalculate the waittime to action repeattime, use smaller value to recheck more frequently
- Allow target SMART_TARGET_CLOSEST_PLAYER to be used by gameobjects
- Allow target SMART_TARGET_OWNER_OR_SUMMONER to be used by gameobjects
- Fixed SMART_EVENT_COUNTER_SET to be only called for the id that was incremented
- Changed the way counters work
- Protect PhaseInc from surpassing maximum phase
- Added loading checks for missing NON_REPEATABLE flag if no repeatmin, repeatmax is set
- Added spell validation for SMART_ACTION_CROSS_CAST

**Target branch(es):** 3.3.5

- [x] 3.3.5

Review and tests required.

### CODE REVIEW

File: src/server/game/Entities/Object/Object.h [hunk 1]
```diff
@@ -557,6 +557,7 @@ class TC_GAME_API WorldObject : public Object, public WorldLocation
         Creature*   FindNearestCreature(uint32 entry, float range, bool alive = true) const;
         GameObject* FindNearestGameObject(uint32 entry, float range) const;
         GameObject* FindNearestGameObjectOfType(GameobjectTypes type, float range) const;
+        Player* WorldObject::SelectNearestPlayer(float distance) const;
```
Comment (Shauren) [inline src/server/game/Entities/Object/Object.h]: wont compile everywhere

Comment (xinef1) [inline src/server/game/Entities/Object/Object.h]: I know, forgot to save the file before commit. Will fix in a moment

**Comment by xinef1** [comment]
```sql

ALTER TABLE `smart_scripts` CHANGE `event_flags` `event_flags` smallint(5) unsigned NOT NULL DEFAULT '0';
ALTER TABLE `smart_scripts` CHANGE `event_phase_mask` `event_phase_mask` smallint(5) unsigned NOT NULL DEFAULT '0';


```

**Comment by Kittnz** [comment]
And this won't break any existing scripts?

**Comment by xinef1** [comment]
Absolutely shouldn't break anything, except for pooped entries in database:
- target 0 in actions which were adjusted to work with targetlist
- action 69 (move to pos) with wrong target type (used coordinates but target is not 8)

Ill create issue with queries to detect some of those failures.

**Comment by Aokromes** [comment]
IMHO you must split added actions/modified actions/extended targets from other fixes.

**Comment by xinef1** [comment]
I wouldn't mind if someone else could split this into separate prs, Ill have less free time starting tomorrow.

**Comment by Palabola** [comment]
> Don't call SMART_EVENT_RESPAWN for dead units before they actually respawn
> Don't call SMART_EVENT_RESPAWN for not spawned gameobjects

**These are more feature than actualy a bug, used several times!**

**Comment by xinef1** [comment]
This a bug and not a feature. If you want to hook creation, use SMART_EVENT_JUST_CREATED or SMART_EVENT_AI_INIT

**Comment by Palabola** [comment]
@xinef1 True, I was wrong just :D Early morning...

**Comment by DDuarte** [comment]
@xinef1: what's your opinion on bfd17566b31bfee91ba51b3cc423f25b588df815?
tl;dr it makes GetTarget return a std::vector instead of a list*

**Comment by xinef1** [comment]
@DDuarte added some comments, using phone :p you may also change objectlist to something reflecting real functionality ObjectVector or just Objects

**Comment by xinef1** [comment]
@DDuarte added some comments, using phone :p you may also change objectlist to something reflecting real functionality ObjectVector or just Objects

File: src/server/game/AI/CoreAI/UnitAI.h [hunk 2]
```diff
@@ -115,6 +115,33 @@ struct TC_GAME_API NonTankTargetSelector : public std::unary_function<Unit*, boo
         bool _playerOnly;
 };
 
+// Simple selector for units using mana
+struct TC_GAME_API PowerUsersSelector : public std::unary_function<Unit*, bool>
```
Comment (Shauren) [inline src/server/game/AI/CoreAI/UnitAI.h]: unary_function is deprecated in C++11.

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 3]
```diff
@@ -1493,6 +1577,7 @@ class TC_GAME_API SmartWaypointMgr
 
 // all events for a single entry
 typedef std::vector<SmartScriptHolder> SmartAIEventList;
+typedef std::list<SmartScriptHolder> SmartAIEventStoredList;
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: list? do you need the list properties?

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 4]
```diff
@@ -2681,7 +2904,7 @@ ObjectList* SmartScript::GetTargets(SmartScriptHolder const& e, Unit* invoker /*
                 l->assign(objectList->begin(), objectList->end());
             }
 
-            return l;
+            break; // previously return l; memory leak
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: it doesnt matter what was here before - unneccessary comment

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 5]
```diff
@@ -2965,11 +3215,15 @@ void SmartScript::ProcessEvent(SmartScriptHolder& e, Unit* unit, uint32 var0, ui
         case SMART_EVENT_JUST_SUMMONED:
         case SMART_EVENT_RESET:
         case SMART_EVENT_JUST_CREATED:
-        case SMART_EVENT_GOSSIP_HELLO:
         case SMART_EVENT_FOLLOW_COMPLETED:
         case SMART_EVENT_ON_SPELLCLICK:
             ProcessAction(e, unit, var0, var1, bvar, spell, gob);
             break;
+        case SMART_EVENT_GOSSIP_HELLO: // added no report use distinction for gameobjects
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: comments like this belong on pr, not in code

**Review by Shauren — CHANGES_REQUESTED** [review]

**Review by Shauren — APPROVED** [review]
